### PR TITLE
MINOR: Enable ignored upgrade system tests - trunk

### PR DESCRIPTION
**Pull request by bbejeck**
Removed `ignore` annotations from the upgrade tests.  This PR includes the following changes for updating the upgrade tests:

1. Uploaded new versions `0.10.2.2`, `0.11.0.3`, `1.0.2`, `1.1.1`, and `2.0.0` (in the associated scala versions) to kafka-packages
2. Update versions in `version.py`, `Dockerfile`, `base.sh`
3. Added new versions to `StreamsUpgradeTest.test_upgrade_downgrade_brokers` *_including_* version `2.0.0` 
4. Added new versions `StreamsUpgradeTest.test_simple_upgrade_downgrade` test *_excluding_* version `2.0.0`
     1. Version `2.0.0` is excluded from the streams upgrade/downgrade test as `StreamsConfig` needs an update for the new version, requiring a KIP. Once the community votes the KIP in, a minor follow-up PR can be pushed to add the `2.0.0` version to the upgrade test.
5. Fixed minor bug in `kafka-run-class.sh` for classpath in upgrade/downgrade tests across versions.

Follow on PRs for `0.10.2x`, `0.11.0x`, `1.0.x`, `1.1.x`, and `2.0.x` will be pushed soon with the same updates required for the specific version.

For testing, the following https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1942/ branch builder was executed and passed.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: bin/kafka-run-class.sh
Comment (bbejeck) [inline bin/kafka-run-class.sh:110]: This change is required so associated `upgrade-system-tests-XX` module is first in the classpath and the updated `StreamsUpgradeTest` for the specific version is run by the upgrade/downgrade tests

File: build.gradle
Comment (bbejeck) [inline build.gradle:1192]: Not needed immediately, but I put this in so when we release `2.1` we are set for 2.0 upgrade/downgrade tests.

**Comment by bbejeck** [comment]
@guozhangwang @mjsax @vvcephei for reviews

File: tests/kafkatest/tests/streams/streams_upgrade_test.py
Comment (guozhangwang) [inline tests/kafkatest/tests/streams/streams_upgrade_test.py]: How about making it just as `metadata_2_versions` minus `str(LATEST_0_10_1)`, and leave a comment that this variable can be completely replaced with `metadata_2_versions` after `0.10.1.2 is available`.

**Review by guozhangwang — COMMENTED** [review]
One minor comment, otherwise lgtm.

**Review by vvcephei — APPROVED** [review]
lgtm. Thanks for doing this!

File: bin/kafka-run-class.sh
Comment (mjsax) [reply bin/kafka-run-class.sh:110]: Cannot follow. Can you elaborate? Also, why was this not an issue in the past?

File: build.gradle
Comment (mjsax) [reply build.gradle:1192]: Why not needed now? We would use this to test upgrading from 2.0 to 2.1-SNAPSHOT

File: streams/upgrade-system-tests-20/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java
Comment (mjsax) [inline streams/upgrade-system-tests-20/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java]: Why? IIRC, we can use this to test upgrading from `2.0` to `2.1-SNAPSHOT` -- we don't need to wait for the release, do we?

File: tests/docker/Dockerfile
Comment (mjsax) [inline tests/docker/Dockerfile]: If we agree to only download the latest bug-fix version, we can remove some lines here (cf. Guozhang's comment on the other PR)

File: tests/kafkatest/tests/streams/streams_upgrade_test.py
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_upgrade_test.py]: Compare my comment on the other PR -- we might want to rewrite to replace `sleep` calls

File: bin/kafka-run-class.sh
Comment (vvcephei) [reply bin/kafka-run-class.sh:110]: Hmm, now that I think about this more, if memory serves, this is not a reliable way to ensure the right class gets loaded, since the classloader spec specifically leaves undefined how a class should be chosen when multiple are available.

Maybe we should just give the classes unique names instead?

Comment (bbejeck) [reply bin/kafka-run-class.sh:110]: That's true extension classes (https://docs.oracle.com/javase/8/docs/technotes/tools/findingclasses.html#extclass)  but for user classes, the first one found on the classpath is what is used

Comment (bbejeck) [reply bin/kafka-run-class.sh:110]: >Cannot follow. Can you elaborate? Also, why was this not an issue in the past?

When searching for classes on the classpath, if multiple classes with the same name exist on the classpath, the first one found wins.  This was causing the failures of the test, as it was using the `StreamsUpgradeTest` found in trunk.

For the `1.1.1` upgrade the test is using the `StreamsUpgradeTest` in trunk and not the `upgrade-system-tests-11` module and failing due changes how the test is loaded.  It wasn't an issue before because we were specifying those versions.  The `StreamsUpgradeTest` in the `upgrade-system-tests` expects different arguments (kafka bootstrap servers string and a properties file name).

I made the changes to the other `kafka-run-class` assuming the same problem would occur as it was the same situation.

Comment (vvcephei) [reply bin/kafka-run-class.sh:110]: Thanks for the reference. It might be a failure to read, but I couldn't find where it specifies that it searches the classpath in order. Only that it searches jar manifests in order.

It seems plausible that the file classloader does actually traverse the classpath in order, but it still looks like it's not guaranteed, since it's not in the spec.

It may not be a huge deal, since the only impact of loading the wrong class is a flaky test, which we can easily just fix again later on.

File: build.gradle
Comment (bbejeck) [reply build.gradle:1192]: Sure, but we can't include `2.0` in the test until we get a KIP for updating `StreamsConfig` for the `2.0` version.

File: tests/kafkatest/tests/streams/streams_upgrade_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_upgrade_test.py]: Not sure I follow, as it's already the same less `str(LATEST_0_10_1)` I'll update the comment as you suggest.

File: streams/upgrade-system-tests-20/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java
Comment (bbejeck) [reply streams/upgrade-system-tests-20/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java]: I can remove the comment, just following the pattern from other upgrade-system-test modules

File: tests/docker/Dockerfile
Comment (bbejeck) [reply tests/docker/Dockerfile]: ack

File: tests/kafkatest/tests/streams/streams_upgrade_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_upgrade_test.py]: maybe we could merge as is and rewrite in a follow-on PR?  WDYT?

**Comment by bbejeck** [comment]
updated this

File: build.gradle
Comment (guozhangwang) [reply build.gradle:1192]: @bbejeck If `StreamsConfig#TOPOLOGY_OPTIMIZATION` is set to `NO_OPTIMIZATION`, which is the default value, it should be compatible still right?

Comment (bbejeck) [reply build.gradle:1192]: Yes, that's correct.

Comment (guozhangwang) [reply build.gradle:1192]: So can we add the test from 2.0 to 2.1-SNAPSHOT in this PR as well?

Comment (bbejeck) [reply build.gradle:1192]: Not yet `StreamsConfigs` checks for a valid version to upgrade from and adding `2.0` fails
https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java#L675

It's my understanding we need a KIP to add `2.0`.  If we don't need a KIP, then I can add `2.0` to `StreamsConfig` and update the test.

Comment (guozhangwang) [reply build.gradle:1192]: We should just be able to use the default value (which is `none`) in 2.1-SNAPSHOT right?

Comment (mjsax) [reply build.gradle:1192]: We don't need to add `2.0` as `upgrade.from` config -- with version probing, upgrading from 2.0 to 2.1 must work without the config specified.

File: tests/kafkatest/tests/streams/streams_upgrade_test.py
Comment (mjsax) [reply tests/kafkatest/tests/streams/streams_upgrade_test.py]: It will result in the same list of versions -- both equally good IMHO.

File: tests/kafkatest/tests/streams/streams_upgrade_test.py
Comment (mjsax) [reply tests/kafkatest/tests/streams/streams_upgrade_test.py]: No strong opinion -- a slight preference to keep it disabled and enable and rewrite in follow up PR. I agree, that we should exclude rewriting the test from this PR.

**Comment by guozhangwang** [comment]
LGTM, cc @mjsax or @vvcephei for another look before merging.

**Review by enothereska — APPROVED** [review]

**Comment by bbejeck** [comment]
retest this please

File: vagrant/base.sh
Comment (vvcephei) [inline vagrant/base.sh]: I think we can also drop the old versions here, to speed up vagrant-based tests.

File: tests/kafkatest/version.py
Comment (vvcephei) [inline tests/kafkatest/version.py:96]: If we're not downloading these old versions anymore, do we need to keep these variables defined? It seems like it wouldn't be possible to actually use them, so we could just delete the definitions to surface the error earlier.

Not sure, though...

**Review by vvcephei — COMMENTED** [review]
Just a couple more questions.

File: vagrant/base.sh
Comment (bbejeck) [reply vagrant/base.sh]: ack

File: tests/kafkatest/version.py
Comment (bbejeck) [reply tests/kafkatest/version.py:96]: I'll leave the old versions here as the version.py is other tests outside streams my need access to older versions.

File: tests/kafkatest/tests/streams/streams_upgrade_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_upgrade_test.py]: I'll put the ignore annotation back until we refactor the test to remove sleep calls

**Comment by bbejeck** [comment]
updated this

**Comment by bbejeck** [comment]
kicked off new branch builder https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1960/

File: tests/kafkatest/version.py
Comment (mjsax) [reply tests/kafkatest/version.py:96]: Agreed with @bbejeck -- we should leave this it. Maintaining those versions is not Streams specific but for all Kafka components.

File: tests/kafkatest/tests/streams/streams_upgrade_test.py
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_upgrade_test.py]: Should this be `[str(LATEST_2_0), str(DEV_VERSION)]` ?

Or actually, `trunk` is already on version 4 -- however, since version 3 we have "version probing" and thus we don't need to distinguish versions 3,4,... -- maybe we can rename the variable to `metadata_3_or_higher_versions` (or a similar but better name) ?

**Review by mjsax — COMMENTED** [review]
One minor comment. Overall LGTM.

File: tests/kafkatest/tests/streams/streams_upgrade_test.py
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_upgrade_test.py]: ack for adding `LATEST_2_0`.  As for the name I can't come up with anything better at the moment so I'll go with what you suggested.
I'll kick off another branch builder

**Comment by bbejeck** [comment]
updated this

**Comment by bbejeck** [comment]
kicked off https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1963/

EDIT: latest branch builder passed

**Comment by bbejeck** [comment]
rebased this to pick up changes for failing tests

**Review by mjsax — APPROVED** [review]
LGTM, assuming all tests pass.

**Comment by bbejeck** [comment]
all tests passing now

**Pull request merged**
